File: src/main/java/com/bioinformatica/bioinformatics/BioInformatics.java
import tech.tablesaw.api.DoubleColumn;
import tech.tablesaw.api.StringColumn;
import tech.tablesaw.api.Table;
import org.RDKit.*;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BioInformatics {

    // Define the delta function
    public static int delta(char x, char y) {
        return (x == y) ? 0 : 1;
    }

    // Define the M function
    public static int M(String seq1, String seq2, int i, int j, int k) {
        int sum = 0;
        for (int index = 0; index < k; index++) {
            sum += delta(seq1.charAt(i + index), seq2.charAt(j + index));
        }
        return sum;
    }

    // Define the makeMatrix function
    public static int[][] makeMatrix(String seq1, String seq2, int k) {
        int n = seq1.length();
        int m = seq2.length();
        int[][] matrix = new int[n - k + 1][m - k + 1];

        for (int i = 0; i < n - k + 1; i++) {
            for (int j = 0; j < m - k + 1; j++) {
                matrix[i][j] = M(seq1, seq2, i, j, k);
            }
        }

        return matrix;
    }

    // Define the plotMatrix function
    public static void plotMatrix(int[][] M, int t, String seq1, String seq2, char nonblank, char blank) {
        System.out.println(" |" + seq2);
        System.out.println("-".repeat(2 + seq2.length()));

        for (int i = 0; i < seq1.length(); i++) {
            System.out.print(seq1.charAt(i) + "|");
            for (int j = 0; j < seq2.length(); j++) {
                char s = M[i][j] < t ? nonblank : blank;
                System.out.print(s);
            }
            System.out.println();
        }
    }

    // Define the dotplot function
    public static void dotplot(String seq1, String seq2, int k, int t) {
        int[][] M = makeMatrix(seq1, seq2, k);
        plotMatrix(M, t, seq1, seq2, '\u25A0', ' ');
    }

    public static double gcContent(String seq) {
        long countG = seq.chars().filter(ch -> ch == 'G').count();
        long countC = seq.chars().filter(ch -> ch == 'C').count();
        return (double) (countG + countC) / seq.length() * 100;
    }

    public static double atContent(String seq) {
        long countA = seq.chars().filter(ch -> ch == 'A').count();
        long countT = seq.chars().filter(ch -> ch == 'T').count();
        return (double) (countA + countT) / seq.length() * 100;
    }

    public static double aromaticProportion(String smiles) {
        try {
            RWMol mol = RWMol.MolFromSmiles(smiles);
            int numAtoms = mol.getNumAtoms();
            int aromaticAtomCount = 0;

            for (int i = 0; i < numAtoms; i++) {
                Atom atom = mol.getAtomWithIdx(i);
                if (atom.getIsAromatic()) {
                    aromaticAtomCount++;
                }
            }

            return (double) aromaticAtomCount / numAtoms;
        } catch (Exception e) {
            e.printStackTrace();
            return 0.0;
        }
    }

    public static Table generate(List<String> smilesList) {
        Table descriptors = Table.create("Molecule Descriptors");

        // Create columns with the same number of rows as smilesList
        StringColumn smilesColumn = StringColumn.create("SMILES");
        DoubleColumn molLogPColumn = DoubleColumn.create("MolLogP");
        DoubleColumn molWtColumn = DoubleColumn.create("MolWt");
        DoubleColumn numRotatableBondsColumn = DoubleColumn.create("NumRotatableBonds");
        DoubleColumn aromaticPropColumn = DoubleColumn.create("AromaticProportion");

        for (String smiles : smilesList) {
            try {
                RWMol mol = RWMol.MolFromSmiles(smiles);
                double molLogP = Descriptors.MolLogP(mol);
                double molWt = Descriptors.MolWt(mol);
                double numRotatableBonds = Descriptors.NumRotatableBonds(mol);
                double aromaticProp = aromaticProportion(smiles);

                // Add data to the columns
                smilesColumn.append(smiles);
                molLogPColumn.append(molLogP);
                molWtColumn.append(molWt);
                numRotatableBondsColumn.append(numRotatableBonds);
                aromaticPropColumn.append(aromaticProp);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        // Add the columns to the table
        descriptors.addColumns(smilesColumn, molLogPColumn, molWtColumn, numRotatableBondsColumn, aromaticPropColumn);

        return descriptors;
    }

    public static Map<Character, Integer> dnaNucleotideCount(String seq) {
        Map<Character, Integer> countMap = new HashMap<>();
        countMap.put('A', 0);
        countMap.put('T', 0);
        countMap.put('G', 0);
        countMap.put('C', 0);

        for (char ch : seq.toCharArray()) {
            countMap.put(ch, countMap.get(ch) + 1);
        }

        return countMap;
    }
}

File: bioinformatica/lib/bioinformatics.dart
import 'dart:math';

int delta(String x, String y) {
  return x == y ? 0 : 1;
}

int M(String seq1, String seq2, int i, int j, int k) {
  int sum = 0;
  for (int index = 0; index < k; index++) {
    sum += delta(seq1[i + index], seq2[j + index]);
  }
  return sum;
}

List<List<int>> makeMatrix(String seq1, String seq2, int k) {
  int n = seq1.length;
  int m = seq2.length;
  List<List<int>> matrix = List.generate(n - k + 1, (i) {
    return List.generate(m - k + 1, (j) {
      return M(seq1, seq2, i, j, k);
    });
  });
  return matrix;
}

void plotMatrix(List<List<int>> M, int t, String seq1, String seq2,
    {String nonblank = '\u25A0', String blank = ' '}) {
  print(' |$seq2');
  print('-' * (2 + seq2.length));
  for (int i = 0; i < seq1.length; i++) {
    String line = seq1[i] +
        '|' +
        M[i].map((s) => s < t ? nonblank : blank).join('');
    print(line);
  }
}

void dotplot(String seq1, String seq2, {int k = 1, int t = 1}) {
  List<List<int>> M = makeMatrix(seq1, seq2, k);
  plotMatrix(M, t, seq1, seq2);
}

double gcContent(String seq) {
  int countG = seq.split('G').length - 1;
  int countC = seq.split('C').length - 1;
  return (countG + countC) / seq.length * 100;
}

double atContent(String seq) {
  int countA = seq.split('A').length - 1;
  int countT = seq.split('T').length - 1;
  return (countA + countT) / seq.length * 100;
}

double aromaticProportion(Chem.Mol mol) {
  List<bool> aromaticAtoms = List.generate(mol.getNumAtoms(), (i) {
    return mol.getAtomWithIdx(i).getIsAromatic();
  });
  int aaCount = aromaticAtoms.where((i) => i == true).length;
  int heavyAtom = Descriptors.HeavyAtomCount(mol);
  double AR = aaCount / heavyAtom;
  return AR;
}

class MolecularDescriptor {
  double molLogP;
  double molWt;
  double numRotatableBonds;
  double aromaticProportion;

  MolecularDescriptor(
      this.molLogP, this.molWt, this.numRotatableBonds, this.aromaticProportion);
}

List<MolecularDescriptor> generate(List<String> smilesList) {
  List<MolecularDescriptor> descriptors = [];

  for (String smiles in smilesList) {
    Chem.Mol? mol = Chem.MolFromSmiles(smiles);
    if (mol != null) {
      double molLogP = Descriptors.MolLogP(mol);
      double molWt = Descriptors.MolWt(mol);
      double numRotatableBonds = Descriptors.NumRotatableBonds(mol);
      double aromaticProp = aromaticProportion(mol);

      MolecularDescriptor descriptor =
          MolecularDescriptor(molLogP, molWt, numRotatableBonds, aromaticProp);
      descriptors.add(descriptor);
    }
  }

  return descriptors;
}

Map<String, int> dnaNucleotideCount(String seq) {
  Map<String, int> countMap = {
    'A': seq.split('A').length - 1,
    'T': seq.split('T').length - 1,
    'G': seq.split('G').length - 1,
    'C': seq.split('C').length - 1,
  };
  return countMap;
}
